src/QsCompiler/Core/BaseTransformation.fs

```diff
@@ -0,0 +1,683 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.Core
+
+open System.Collections.Immutable
+open System.Linq
+open System.Numerics
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+type QsArgumentTuple = QsTuple<LocalVariableDeclaration<QsLocalSymbol>>
+type private ExpressionKind = QsExpressionKind<TypedExpression,Identifier,ResolvedType>
+type private ExpressionType = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>
+
+type MyHandler<'T> (handler : 'T -> 'T) = member this.Call = handler
+
+type MyEvent<'T,'F> (defaultHandler : 'T -> 'T, final : 'T -> 'F) =
+    
+    let removeAt i lst =
+        List.append
+        <| List.take i lst
+        <| List.skip (i+1) lst
+
+    member val private Handlers : MyHandler<'T> list = [] with get, set
+    member val private CombinedHandleValid = true with get, set
+    member val private CombinedHandle = defaultHandler with get, set
+
+    member private this.UpdateCombinedHandle () =
+        this.CombinedHandle <- match this.Handlers with
+                               | [f] -> f.Call
+                               | f1 :: fn -> List.fold (>>) f1.Call <| List.map (fun (f : MyHandler<'T>) -> f.Call) fn
+                               | [] -> defaultHandler
+        this.CombinedHandleValid <- true
```

Comment:
Is the added complexity of caching a combined handler worth it? It seems that you can avoid traversing the list each time, but you still have to call all of the functions in sequence in both cases. Is there a performance difference?